docs/spec/identities.md

```diff
@@ -0,0 +1,427 @@
+---
+title: "Radicle Link -- Section: Identities"
+author:
+- Kim Altintop[^0]
+status: draft
+date: 2020-07-15
+
+numbersections: true
+link-citations: true
+bibliography: references.bib
+reference-links: true
+...
+[^0]: [kim@monadic.xyz](mailto:kim@monadic.xyz)
+
+# Identities
+
+In order to collaborate on repositories within a consensus-free network, we must
+be able to refer to them using a stable identifier. Note that this identifier is
+a **statement of intent**: a repository can be described as a collection of
+ever-moving leaves of a DAG whose root element is the empty object. Therefore,
+the _content_ of a repository is not enough to describe it -- while two views on
+the repository may share objects, they may diverge substantially otherwise. Both
+views may however state their intent to eventually converge to the same state.
+
+While in principle a random identifier with sufficient entropy would suffice for
+the purpose, this would put the burden of deciding which repository views are
+_legit_ (for some definition of legit) entirely on the user. Instead, our
+approach is to establish an ownership proof, tied to the network identity of a
+peer (`PeerId`), or a set of peers, such that repository views can be replicated
+according to the trust relationships between peers ("tracking").
+
+Our model is based on The Update Framework (TUF)[@tuf], conceived as a means of
+securely distributing software packages.
+
+## Data Model
+
+The identity of a repository is established using a document of the form:
+
+```rust
+struct Doc<T, D> {
+    replaces: Option<Revision>,
+    payload: T,
+    delegations: D,
+}
+```
+
+where:
+
+* `Revision` is a cryptographic hash of a document's contents, such that this
+  document is content-addressable by this hash within the storage system.
+* `replaces` refers to the previous revision of the document, or none if it is
+  the first revision
+* `payload` is an extensible, forwards- and backwards-compatible datatype
+  containing application-defined metadata about the repository. The protocol
+  interprets _some_ of the properties, as described in [TODO].
+* `delegations` contains the public keys of key owners who are authorised to
+  issue and approve new revisions of the document. The delegation format depends
+  on the type of identity being established, as detailed below.
+
+The `Doc` MUST be serialised in canonical form, e.g. [Canonical JSON].
+
+The authenticity of the `Doc` is captured by the following type:
+
+```rust
+struct Identity<T, D> {
+    id: ContentId,
+    root: Revision,
+    revision: Revision,
+    doc: Doc<T, D>,
+    signatures: HashMap<PublicKey, Signature>,
+}
+```
+
+where:
+
+* `id` is the content-addressable hash of the `Identity` itself
```

Comment:
What is the significance of this?

Reply:
I believe the idea is that we should be able to fetch data by its hash, whether it's a `Doc` or and `Identity`. So the `id` here is the fetchable hash for `Identity`. In the **Git** section you can see that the `id` is the SHA identifier of the commit.

Reply:
Okay, so doesn't really play a role in the protocol?

Reply:
I think it means if we picked another backing store, let's say sqlite, we would store these data inside using their hashes as the key.

That's what I gather, at least.

Reply:
There can be more than one `Identity` (= commit) which verifies a given `Doc` history (= tree + parent trees), so we SHOULD be able to retrieve the ones we already validated (by doc revision). This is significant in a system where consensus is a social construct — although it makes things less confusing if you assign a “merge manager” and keep both histories mostly aligned (as detailed in the “Workflow” section).

Reply:
How do we validate that a cached validation wasn't fabricated?

Reply:
I think if we turned the question on its head we might get an answer: Why would you cache a non-verified identity?

Reply:
Pre-validated identities are obviously _not_ shared on the network. If you don't trust your own storage backend, you might consider to sign the validated-markers, but I'm not sure that'd be worth the trouble.

Reply:
I was thinking more of a bad actor violating the cache to bypass some validation. But it's clear now.